### ISSUE TITLE
Several errors, when connecting to HPE iLO5

### ISSUE THREAD

**Issue by daniel-wtd** [ManageIQ/manageiq-providers-redfish#89]
Hi y'all,

first let me thank you for the awesome work.
Over the last days we were trying to get the PXE Deployment running with the redfish provider as described here (https://steampunk.si/posts/miq-bare-metal-provisioning/)

Unfortunately, we are running into multiple issues.

Information:

- manageiq hammer-9 (8 has the same issues)
- server is an HPE with iLO5
- redfish provider and pxe provider involved

1. redfish auth works
2. Resource scanning / refresh takes hours
3. Errors are visible in the refish provider webfrontend (example):
```
Error - About 2 Hours Ago
--
undefined method `Members' for nil:NilClass
```
4. Additional errors are visible in the redfish.log:

```
# spam / multiple of
[----] W, [2019-08-12T07:44:26.448384 #23520:bfc0d54]  WARN -- : Disconnect failed: RedfishClient::Resource::NoODataId
```
5. Some more errors occure in the evm.log
```
[----] E, [2019-08-12T07:46:48.085088 #23520:4de6bd4] ERROR -- : /usr/local/lib/ruby/gems/2.4.0/bundler/gems/manageiq-providers-redfish-17a53bcf5bf7/app/models/manageiq/providers/redfish/physical_infra_manager/event_catcher/runner.rb:5:in `monitor_events'
[----] E, [2019-08-12T07:46:48.085018 #23520:4de6bd4] ERROR -- : MIQ(ManageIQ::Providers::Redfish::PhysicalInfraManager::EventCatcher::Runner#start_event_monitor) EMS [10.21.254.18] as [proficom] Event Monitor Thread aborted because [undefined method `listen' for nil:NilClass]
[----] E, [2019-08-12T04:13:04.573789 #372:f48f58] ERROR -- : /usr/local/lib/ruby/gems/2.4.0/bundler/gems/manageiq-providers-redfish-17a53bcf5bf7/app/models/manageiq/providers/redfish/inventory/parser/physical_infra_manager.rb:90:in `block in get_server_cpu_core_count'
[----] E, [2019-08-12T04:58:04.843150 #18469:f48f58] ERROR -- : /usr/local/lib/ruby/gems/2.4.0/bundler/gems/manageiq-providers-redfish-17a53bcf5bf7/app/models/manageiq/providers/redfish/inventory/collector/physical_infra_manager.rb:4:in `physical_servers'
[----] E, [2019-08-12T06:13:29.896683 #29135:f48f58] ERROR -- : MIQ(ManageIQ::Providers::Redfish::PhysicalInfraManager::Refresher#refresh) EMS: [hp01], id: [4] Refresh failed
[----] E, [2019-08-12T06:58:26.055676 #15066:f48f58] ERROR -- : /usr/local/lib/ruby/gems/2.4.0/bundler/gems/manageiq-providers-redfish-17a53bcf5bf7/app/models/manageiq/providers/redfish/inventory/collector/physical_infra_manager.rb:8:in `physical_racks'
```
6. The new servers are not deployable (via lifecycle), because we cannot choose a method for them. (see screenshots attached)

![Screenshot from 2019-08-12 09-51-33](https://user-images.githubusercontent.com/6490538/62851703-499cf080-bce7-11e9-8f38-f2d4c7d35a7e.png)
![Screenshot from 2019-08-12 09-53-24](https://user-images.githubusercontent.com/6490538/62851705-499cf080-bce7-11e9-95e3-07fa6e134033.png)

For now, we can get all the required information via ansible and expect some other issue. According to the errors, we are a little bit puzzled how to continue here.

It would be awesome to get some advice or help.
If you need any other information, please feel free to reach out to me.

Thank you so much in advance for your awesome work.

**Comment by daniel-wtd** [comment]
Small addition, deployment is also not choosable with the mock, described here: https://steampunk.si/posts/manageiq-redfish/

**Comment by miha-plesko** [comment]
Hi @daniel-wtd, physical server provisioning was massively updated in Ivanchuk release. The form you see when you click `Lifecycle > Provision Physical Server` now looks something like

![Capture](https://user-images.githubusercontent.com/8102426/62853107-9f739780-bceb-11e9-917b-2d23848f63f0.PNG)

I think the Redfish inventory was not really solid yet in Hammer. In other words, you need to switch to Ivanchuk, right @tadeboro ?

**Comment by daniel-wtd** [comment]
I will give it a try.

**Comment by tadeboro** [comment]
MIQ Hammer does not support physical server provisioning at all. The dialog that you see is hardcoded to the Lenovo provider. As @miha-plesko already said, we added support for Redfish-based physical server provisioning in the upcoming Ivanchuk release.

Event catcher is failing because iLO5 does not support SSE part of the Redfish standard. But this should not affect other aspects of the provider.

As for the inventory refresh error, it looks like iLO5 is not providing some data that is mandatory by the Redfish standard. But there is no way of knowing what resource is incomplete without a stack trace that should be present in the logs.

**Comment by daniel-wtd** [comment]
Thanks a lot. I am spining up the Ivanchuk in some minutes, have a look and will come back asap.

**Comment by daniel-wtd** [comment]
Hi,
so far so good. I was able to get "some" functionality. Thank you for the awesome support! This is teriffic.

1. Able to authenticate
2. Able to choose provision
3. Able to choose my pxe-server, image and customization
4. Server remains off
5. I cannot see any customization files or changes on my pxe server :/

The log is somewhat silent regarding the creation of files or steps after clickling the "submit" during provision. I can see my interactions in the production.log.

Any hints where I can have a look at? For now, I can start/stop/reboot the system via redfish interactions, but provision seems to do nothing.

**Comment by miha-plesko** [comment]
@daniel-wtd did you Approve provisioning request? You have to go to Services > Requests, open the provisioning request and approve it. Until you do this, nothing will happen.

Once approved, I would check automation.log. You should see following steps there:
```
start_provisioning
deploy_pxe_config # this is where PXE is supposed to get configured and server set to one-time PXE boot
poll_server_on_initial
power_off_server
poll_server_off
power_on_server
poll_server_on
```

**Comment by matejart** [comment]
Out of curiosity, do the basic operations of powering on and off from the ManageIQ web interface work? If you have physical access to the system, you'll see if there is any effect within a few seconds. If not, please either use either iLO's web interface or perform *Configuration* -> *Refresh Relationships and Power States*.

Any errors will appear in `emv.log`, but in this case please paste the whole stack.

**Comment by daniel-wtd** [comment]
@matejart yep, they work (almost every time)

@miha-plesko That was the trick, now coming to the next issue ^^

```
 [EVM] PhysicalServer [1|/redfish/v1/Systems/1] Step [CheckProvision] Message [[NoMethodError]: undefined method `MACAddress' for nil:NilClass] 
```

**Comment by daniel-wtd** [comment]
Via Ansible, we can get the macaddresses properly for all (6) interfaces. :/                                                                                                                                                                                                      
```
            "LinkStatus": "LinkUp",                                                                                                                                                                                                        
            "MACAddress": "f4:03:43:c2:da:40",
```

**Comment by tadeboro** [comment]
This is strange since both the Ansible module and MIQ use data from the same source.  In what state was the server (powered on, powered off) when you tried to provision it and when you retrieved the facts using Ansible? HPE servers tend to be a bit picky about when the network data is available over Redfish.

**Comment by daniel-wtd** [comment]
Thats a good question, maybe I can try a refresh in different states.

**Comment by tadeboro** [comment]
Another possible scenario in which this could happen is if the iLO fails to respond to one of the `GET` queries that fetch NIC details.

**Comment by daniel-wtd** [comment]
refreshing in several states (on, off, post) lead to no conclusion from my side.

I had a look in the logs related to MAC or mac_address:

```
automation.log:[----] E, [2019-08-12T08:01:33.073205 #2792:a52588c] ERROR -- : Q-task_id([r13_physical_server_provision_task_1]) <AEMethod update_provision_status> PhysicalServer Provision Error: [EVM] PhysicalServer [1|/redfish/v1/Systems/1] Step [CheckProvision] Message [[NoMethodError]: undefined method `MACAddress' for nil:NilClass] 
automation.log:[----] E, [2019-08-12T08:12:20.720710 #2792:d9b2e10] ERROR -- : Q-task_id([r14_physical_server_provision_task_2]) <AEMethod check_provision> ProvisionCheck (PhysicalServer id=1 ems_ref=/redfish/v1/Systems/1) error <[NoMethodError]: undefined method `MACAddress' for nil:NilClass>
```

Otherwise it is quite hard to find something.

**Comment by tadeboro** [comment]
This error indicates that when our Redfish client iterated over the NICs, iLO failed t return the valid data. My guess would be that the BMC was under some stress when you were trying to execute the provisioning process and had a minor "stroke" somewhere in the middle of the operation.

How many times did you try to provision the system? And is the error reproducible?

**Comment by daniel-wtd** [comment]
I have rebootet the manageiq ~15 minutes ago and was trying to provision only one:

the log shows this behaviour now:

```
[----] I, [2019-08-12T10:17:15.426336 #2792:aa45ec]  INFO -- : Q-task_id([r16_physical_server_provision_task_4]) <AEMethod [/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Methods/check_provision]> Starting 
[----] I, [2019-08-12T10:17:15.677192 #2792:ea3fcec]  INFO -- : Q-task_id([r16_physical_server_provision_task_4]) <AEMethod check_provision> ProvisionCheck (PhysicalServer id=1 ems_ref=/redfish/v1/Systems/1) returned <error> for state <finished> and status <Error>
[----] E, [2019-08-12T10:17:15.679145 #2792:ea3fcec] ERROR -- : Q-task_id([r16_physical_server_provision_task_4]) <AEMethod check_provision> ProvisionCheck (PhysicalServer id=1 ems_ref=/redfish/v1/Systems/1) error <[NoMethodError]: undefined method `MACAddress' for nil:NilClass>
[----] I, [2019-08-12T10:17:15.696478 #2792:aa45ec]  INFO -- : Q-task_id([r16_physical_server_provision_task_4]) <AEMethod [/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Methods/check_provision]> Ending
[----] I, [2019-08-12T10:17:15.696546 #2792:aa45ec]  INFO -- : Q-task_id([r16_physical_server_provision_task_4]) Method exited with rc=MIQ_OK
[----] I, [2019-08-12T10:17:15.696771 #2792:aa45ec]  INFO -- : Q-task_id([r16_physical_server_provision_task_4]) Followed  Relationship [miqaedb:/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Methods/CheckProvision#redfish_ph_infra]
[----] I, [2019-08-12T10:17:15.696819 #2792:aa45ec]  INFO -- : Q-task_id([r16_physical_server_provision_task_4]) Processed State=[CheckProvision] with Result=[error]
[----] W, [2019-08-12T10:17:15.696852 #2792:aa45ec]  WARN -- : Q-task_id([r16_physical_server_provision_task_4]) Error in State=[CheckProvision]
[----] I, [2019-08-12T10:17:15.696901 #2792:aa45ec]  INFO -- : Q-task_id([r16_physical_server_provision_task_4]) In State=[CheckProvision], invoking [on_error] method=[update_provision_status(status => "Provisioning Errored")]
[----] I, [2019-08-12T10:17:15.698635 #2792:aa45ec]  INFO -- : Q-task_id([r16_physical_server_provision_task_4]) Updated namespace [PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Provision/update_provision_status  ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines]
[----] I, [2019-08-12T10:17:15.698765 #2792:aa45ec]  INFO -- : Q-task_id([r16_physical_server_provision_task_4]) Invoking [inline] method [/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Provision/update_provision_status] with inputs [{"status"=>"Provisioning Errored"}]
[----] I, [2019-08-12T10:17:15.699563 #2792:aa45ec]  INFO -- : Q-task_id([r16_physical_server_provision_task_4]) <AEMethod [/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Provision/update_provision_status]> Starting 
[----] I, [2019-08-12T10:17:15.964206 #2792:ea875c4]  INFO -- : Validating Notification type: automate_user_error
[----] I, [2019-08-12T10:17:15.965141 #2792:ea875c4]  INFO -- : Calling Create Notification type: automate_user_error subject type: MiqRequest id: 16 options: {:message=>"Physical Server Provision Error: [EVM] PhysicalServer [1|/redfish/v1/Systems/1] Step [CheckProvision] Message [[NoMethodError]: undefined method `MACAddress' for nil:NilClass] "}
[----] E, [2019-08-12T10:17:15.992909 #2792:ea875c4] ERROR -- : Q-task_id([r16_physical_server_provision_task_4]) <AEMethod update_provision_status> PhysicalServer Provision Error: [EVM] PhysicalServer [1|/redfish/v1/Systems/1] Step [CheckProvision] Message [[NoMethodError]: undefined method `MACAddress' for nil:NilClass] 
[----] I, [2019-08-12T10:17:16.009830 #2792:aa45ec]  INFO -- : Q-task_id([r16_physical_server_provision_task_4]) <AEMethod [/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Provision/update_provision_status]> Ending
[----] I, [2019-08-12T10:17:16.009898 #2792:aa45ec]  INFO -- : Q-task_id([r16_physical_server_provision_task_4]) Method exited with rc=MIQ_OK
[----] I, [2019-08-12T10:17:16.010183 #2792:aa45ec]  INFO -- : Q-task_id([r16_physical_server_provision_task_4]) Followed  Relationship [miqaedb:/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Provision/default#create]
[----] I, [2019-08-12T10:17:16.010385 #2792:aa45ec]  INFO -- : Q-task_id([r16_physical_server_provision_task_4]) Followed  Relationship [miqaedb:/physicalinfrastructure/PhysicalServer/Lifecycle/Provisioning#create]
```

**Comment by tadeboro** [comment]
Time to bring out the heavy guns. What does this sequence of commands print out:

    $ BMC='https://bmc.address:port'
    $ CREDS='username:password'
    $ for nic in $(curl -ku $CREDS $BMC/redfish/v1/Systems/1/EthernetInterfaces | jq -r '.Members[] | ."@odata.id"'); do echo $nic; curl -ku $CREDS $BMC$nic; echo; done

The last loop simulates what the MIQ does when trying to retrieve MAC addresses for the PXE menus.

**Comment by daniel-wtd** [comment]
Sorry for the delay @tadeboro ,

1. I have tested the scenario

```
curl -ku "$CREDS" "$BMC/redfish/v1/Systems/1/EthernetInterfaces"

{"@odata.context":"/redfish/v1/$metadata#EthernetInterfaceCollection.EthernetInterfaceCollection","@odata.etag":"W/\"D5EC731D\"","@odata.id":"/redfish/v1/Systems/1/EthernetInterfaces","@odata.type":"#EthernetInterfaceCollection.EthernetInterfaceCollection","Description":"Collection of System Ethernet Interfaces","Name":"System Ethernet Interfaces","Members":[{"@odata.id":"/redfish/v1/Systems/1/EthernetInterfaces/1"},{"@odata.id":"/redfish/v1/Systems/1/EthernetInterfaces/2"},{"@odata.id":"/redfish/v1/Systems/1/EthernetInterfaces/3"},{"@odata.id":"/redfish/v1/Systems/1/EthernetInterfaces/4"},{"@odata.id":"/redfish/v1/Systems/1/EthernetInterfaces/5"},{"@odata.id":"/redfish/v1/Systems/1/EthernetInterfaces/6"}],"Members@odata.count":6}
```

```
curl -ku "$CREDS" "$BMC/redfish/v1/Systems/1/EthernetInterfaces" | jq -r '.Members[] | ."@odata.id"'

/redfish/v1/Systems/1/EthernetInterfaces/1
/redfish/v1/Systems/1/EthernetInterfaces/2
/redfish/v1/Systems/1/EthernetInterfaces/3
/redfish/v1/Systems/1/EthernetInterfaces/4
/redfish/v1/Systems/1/EthernetInterfaces/5
/redfish/v1/Systems/1/EthernetInterfaces/6
```

The interesting part is here:

Result 1, which occurs sometimes:
```
for nic in $(curl -ku "$CREDS" "$BMC/redfish/v1/Systems/1/EthernetInterfaces" | jq -r '.Members[] | ."@odata.id"'); do echo $nic; curl -ku $CREDS $BMC$nic; done

/redfish/v1/Systems/1/EthernetInterfaces/1
{"@odata.context":"/redfish/v1/$metadata#EthernetInterface.EthernetInterface","@odata.etag":"W/\"110F71E3\"","@odata.id":"/redfish/v1/Systems/1/EthernetInterfaces/1","@odata.type":"#EthernetInterface.v1_4_1.EthernetInterface","Id":"1","FullDuplex":true,"IPv4Addresses":[],"IPv4StaticAddresses":[],"IPv6AddressPolicyTable":[],"IPv6Addresses":[],"IPv6StaticAddresses":[],"IPv6StaticDefaultGateways":[],"LinkStatus":"LinkUp","MACAddress":"f4:03:43:c2:da:40","Name":"","NameServers":[],"SpeedMbps":null,"StaticNameServers":[],"Status":{"Health":"OK","State":"Enabled"},"UefiDevicePath":"PciRoot(0x3)/Pci(0x2,0x0)/Pci(0x0,0x0)"}
/redfish/v1/Systems/1/EthernetInterfaces/2
{"@odata.context":"/redfish/v1/$metadata#EthernetInterface.EthernetInterface","@odata.etag":"W/\"C92D5BA9\"","@odata.id":"/redfish/v1/Systems/1/EthernetInterfaces/2","@odata.type":"#EthernetInterface.v1_4_1.EthernetInterface","Id":"2","FullDuplex":true,"IPv4Addresses":[],"IPv4StaticAddresses":[],"IPv6AddressPolicyTable":[],"IPv6Addresses":[],"IPv6StaticAddresses":[],"IPv6StaticDefaultGateways":[],"LinkStatus":"LinkUp","MACAddress":"f4:03:43:c2:da:48","Name":"","NameServers":[],"SpeedMbps":null,"StaticNameServers":[],"Status":{"Health":"OK","State":"Enabled"},"UefiDevicePath":"PciRoot(0x3)/Pci(0x2,0x0)/Pci(0x0,0x0)"}
/redfish/v1/Systems/1/EthernetInterfaces/3
{"@odata.context":"/redfish/v1/$metadata#EthernetInterface.EthernetInterface","@odata.etag":"W/\"78FA3044\"","@odata.id":"/redfish/v1/Systems/1/EthernetInterfaces/3","@odata.type":"#EthernetInterface.v1_4_1.EthernetInterface","Id":"3","FullDuplex":true,"IPv4Addresses":[],"IPv4StaticAddresses":[],"IPv6AddressPolicyTable":[],"IPv6Addresses":[],"IPv6StaticAddresses":[],"IPv6StaticDefaultGateways":[],"LinkStatus":null,"MACAddress":"08:f1:ea:98:b1:0a","Name":"","NameServers":[],"SpeedMbps":null,"StaticNameServers":[],"Status":{"Health":null,"State":null},"UefiDevicePath":"PciRoot(0x0)/Pci(0x1C,0x0)/Pci(0x0,0x0)"}
/redfish/v1/Systems/1/EthernetInterfaces/4
{"@odata.context":"/redfish/v1/$metadata#EthernetInterface.EthernetInterface","@odata.etag":"W/\"A22CE099\"","@odata.id":"/redfish/v1/Systems/1/EthernetInterfaces/4","@odata.type":"#EthernetInterface.v1_4_1.EthernetInterface","Id":"4","FullDuplex":true,"IPv4Addresses":[],"IPv4StaticAddresses":[],"IPv6AddressPolicyTable":[],"IPv6Addresses":[],"IPv6StaticAddresses":[],"IPv6StaticDefaultGateways":[],"LinkStatus":null,"MACAddress":"08:f1:ea:98:b1:0b","Name":"","NameServers":[],"SpeedMbps":null,"StaticNameServers":[],"Status":{"Health":null,"State":null},"UefiDevicePath":"PciRoot(0x0)/Pci(0x1C,0x0)/Pci(0x0,0x0)"}
/redfish/v1/Systems/1/EthernetInterfaces/5
{"@odata.context":"/redfish/v1/$metadata#EthernetInterface.EthernetInterface","@odata.etag":"W/\"5D4E52ED\"","@odata.id":"/redfish/v1/Systems/1/EthernetInterfaces/5","@odata.type":"#EthernetInterface.v1_4_1.EthernetInterface","Id":"5","FullDuplex":true,"IPv4Addresses":[],"IPv4StaticAddresses":[],"IPv6AddressPolicyTable":[],"IPv6Addresses":[],"IPv6StaticAddresses":[],"IPv6StaticDefaultGateways":[],"LinkStatus":null,"MACAddress":"08:f1:ea:98:b1:0c","Name":"","NameServers":[],"SpeedMbps":null,"StaticNameServers":[],"Status":{"Health":null,"State":null},"UefiDevicePath":"PciRoot(0x0)/Pci(0x1C,0x0)/Pci(0x0,0x0)"}

/redfish/v1/Systems/1/EthernetInterfaces/6
{"@odata.context":"/redfish/v1/$metadata#EthernetInterface.EthernetInterface","@odata.etag":"W/\"CCF04762\"","@odata.id":"/redfish/v1/Systems/1/EthernetInterfaces/6","@odata.type":"#EthernetInterface.v1_4_1.EthernetInterface","Id":"6","FullDuplex":true,"IPv4Addresses":[],"IPv4StaticAddresses":[],"IPv6AddressPolicyTable":[],"IPv6Addresses":[],"IPv6StaticAddresses":[],"IPv6StaticDefaultGateways":[],"LinkStatus":null,"MACAddress":"08:f1:ea:98:b1:0d","Name":"","NameServers":[],"SpeedMbps":null,"StaticNameServers":[],"Status":{"Health":null,"State":null},"UefiDevicePath":"PciRoot(0x0)/Pci(0x1C,0x0)/Pci(0x0,0x0)"}
```

Result 2, which was visible in ~10% of the cases:
```
jq: error (at <stdin>:0): Cannot iterate over null (null)
``

So it seems, that the first curl or maybe both are somewhat an issue and can deliver unexpected results. I was not able to replicate this in all cases

2. logwise (after pressing a refresh for the provider

```
# EVM log

evm.log:[----] E, [2019-08-14T10:47:15.782499 #2800:f305fc] ERROR -- : MIQ(ManageIQ::Providers::Redfish::PhysicalInfraManager::Refresher#refresh) EMS: [hp18], id: [2] Refresh failed
evm.log:[----] E, [2019-08-14T10:47:15.782613 #2800:f305fc] ERROR -- : [NoMethodError]: undefined method `dig' for nil:NilClass  Method:[block (2 levels) in <class:LogProxy>]
evm.log:[----] E, [2019-08-14T10:47:15.782666 #2800:f305fc] ERROR -- : /usr/local/lib/ruby/gems/2.5.0/bundler/gems/manageiq-providers-redfish-c8c92b27d884/app/models/manageiq/providers/redfish/inventory/parser/physical_infra_manager.rb:17:in `block in physical_servers'
evm.log:[----] E, [2019-08-14T10:47:15.782702 #2800:f305fc] ERROR -- : MIQ(ManageIQ::Providers::Redfish::PhysicalInfraManager::Refresher#refresh) EMS: [hp18], id: [2] Unable to perform refresh for the following targets:
evm.log:[----] E, [2019-08-14T10:47:15.782758 #2800:f305fc] ERROR -- : MIQ(ManageIQ::Providers::Redfish::PhysicalInfraManager::Refresher#refresh)  --- ManageIQ::Providers::Redfish::PhysicalInfraManager [hp18] id [2]
evm.log:[----] E, [2019-08-14T10:47:15.793704 #2800:f305fc] ERROR -- : MIQ(MiqQueue#deliver) Message id: [9788], Error: [undefined method `dig' for nil:NilClass]
evm.log:[----] E, [2019-08-14T10:47:15.793814 #2800:f305fc] ERROR -- : [ManageIQ::Providers::BaseManager::Refresher::PartialRefreshError]: undefined method `dig' for nil:NilClass  Method:[block (2 levels) in <class:LogProxy>]
evm.log:[----] E, [2019-08-14T10:47:15.793905 #2800:f305fc] ERROR -- : /var/www/miq/vmdb/app/models/manageiq/providers/base_manager/refresher.rb:67:in `refresh'
evm.log:[----] E, [2019-08-14T10:49:42.058461 #2800:f305fc] ERROR -- : MIQ(ManageIQ::Providers::Redfish::PhysicalInfraManager::Refresher#refresh) EMS: [hp18], id: [2] Refresh failed
evm.log:[----] E, [2019-08-14T10:49:42.058622 #2800:f305fc] ERROR -- : [NoMethodError]: undefined method `Links' for nil:NilClass  Method:[block (2 levels) in <class:LogProxy>]
evm.log:[----] E, [2019-08-14T10:49:42.058677 #2800:f305fc] ERROR -- : /usr/local/lib/ruby/gems/2.5.0/bundler/gems/manageiq-providers-redfish-c8c92b27d884/app/models/manageiq/providers/redfish/inventory/parser/physical_infra_manager.rb:100:in `block (2 levels) in hardwares'
evm.log:[----] E, [2019-08-14T10:49:42.058725 #2800:f305fc] ERROR -- : MIQ(ManageIQ::Providers::Redfish::PhysicalInfraManager::Refresher#refresh) EMS: [hp18], id: [2] Unable to perform refresh for the following targets:
evm.log:[----] E, [2019-08-14T10:49:42.058813 #2800:f305fc] ERROR -- : MIQ(ManageIQ::Providers::Redfish::PhysicalInfraManager::Refresher#refresh)  --- ManageIQ::Providers::Redfish::PhysicalInfraManager [hp18] id [2]
evm.log:[----] E, [2019-08-14T10:49:42.069253 #2800:f305fc] ERROR -- : MIQ(MiqQueue#deliver) Message id: [9793], Error: [undefined method `Links' for nil:NilClass]
evm.log:[----] E, [2019-08-14T10:49:42.069364 #2800:f305fc] ERROR -- : [ManageIQ::Providers::BaseManager::Refresher::PartialRefreshError]: undefined method `Links' for nil:NilClass  Method:[block (2 levels) in <class:LogProxy>]
evm.log:[----] E, [2019-08-14T10:49:42.069467 #2800:f305fc] ERROR -- : /var/www/miq/vmdb/app/models/manageiq/providers/base_manager/refresher.rb:67:in `refresh'

[----] E, [2019-08-14T10:52:45.139462 #2790:6ec4bd0] ERROR -- : MIQ(ManageIQ::Providers::Redfish::PhysicalInfraManager::EventCatcher::Runner#start_event_monitor) EMS [10.21.254.18] as [proficom] Event Monitor Thread aborted because [undefined method `listen' for nil:NilClass]
[----] E, [2019-08-14T10:52:45.139564 #2790:6ec4bd0] ERROR -- : [NoMethodError]: undefined method `listen' for nil:NilClass  Method:[block (2 levels) in <class:LogProxy>]
[----] E, [2019-08-14T10:52:45.139590 #2790:6ec4bd0] ERROR -- : /usr/local/lib/ruby/gems/2.5.0/bundler/gems/manageiq-providers-redfish-c8c92b27d884/app/models/manageiq/providers/redfish/physical_infra_manager/event_catcher/runner.rb:5:in `monitor_events'
```

The macaddress(s) do not appear in the webinterface for the server 
![Screenshot from 2019-08-14 12-57-44](https://user-images.githubusercontent.com/6490538/63016162-3595f200-be93-11e9-99ca-79bb91de8cc4.png)

**Comment by tadeboro** [comment]
The `jq` and refresh errors that you see have the exact same origin: the server's BMC (iLO in this particular case) failed to return back a valid JSON response. And if the error rate is anywhere near the 10% you reported, the system is basically useless. Redfish, being OData based API, is a very chatty beast and you need at least five requests to get the NIC data:

 1. `GET /redfish/v1/`
 2. `GET /redfish/v1/Systems/`
 3. `GET /redfish/v1/Systems/1/`
 4. `GET /redfish/v1/Systems/1/EthernetInterfaces/`
 5. `GET /redfish/v1/Systems/1/EthernetInterfaces/1`

The probability that all five requests succeed given the error rate of say 10% is 0.9^5 ~ 60%. I think we can all agree that this is not a good result. When performing an inventory refresh, ManageIQ will send about 30 requests. I will leave the success rate calculation as an exercise for the reader (hint: 0 is a phenomenally good guess).

There is nothing we can do about the flaky BMC software from the ManageIQ's side. Maybe you can try and update the iLO firmware see if that makes the server responses more reliable? If the upgrade does not solve your problems, you will have to talk to the HPE support about the issue you are experiencing.

**Comment by daniel-wtd** [comment]
Thanks a lot @tadeboro,

I will see what I can do here. Maybe there is something like a build in flood detection or whatever HP is doing here. :+1: 

It would be awesome, if we can have the issue open for some more days, so that I can have a chance to see any other solution.

**Comment by daniel-wtd** [comment]
Found something interesting. It seems like the session is dropping somehow or delayed.

```
/redfish/v1/Systems/1/EthernetInterfaces/3
{"error":{"code":"iLO.0.10.ExtendedInfo","message":"See @Message.ExtendedInfo for more information.","@Message.ExtendedInfo":[{"MessageId":"Base.1.0.NoValidSession"}]}}
```

**Comment by daniel-wtd** [comment]
Minor addition: The error rate is at 5-10% when manageiq is connected. if the server is removed from manageiq, I can run my curls without any problem.

After adding a new server, I will see loooots of these errors:

```
[----] I, [2019-08-15T07:12:11.479596 #8541:f305fc]  INFO -- : MIQ(ManageIQ::Providers::Redfish::PhysicalInfraManager::EventCatcher::Runner#start_event_monitor) EMS [10.21.254.18] as [manageiq] Started Event Monitor Thread
[----] I, [2019-08-15T07:12:12.481120 #8541:f305fc]  INFO -- : MIQ(ManageIQ::Providers::Redfish::PhysicalInfraManager::EventCatcher::Runner#do_work) EMS [10.21.254.18] as [manageiq] Event Monitor Thread gone. Restarting...
[----] I, [2019-08-15T07:12:12.481327 #8541:f305fc]  INFO -- : MIQ(ManageIQ::Providers::Redfish::PhysicalInfraManager::EventCatcher::Runner#start_event_monitor) EMS [10.21.254.18] as [manageiq] Validating Connection/Credentials
[----] I, [2019-08-15T07:12:12.481372 #8541:f305fc]  INFO -- : MIQ(ManageIQ::Providers::Redfish::PhysicalInfraManager#with_provider_connection) Connecting through ManageIQ::Providers::Redfish::PhysicalInfraManager: [hp01]
[----] I, [2019-08-15T07:12:12.920818 #8541:f305fc]  INFO -- : MIQ(ManageIQ::Providers::Redfish::PhysicalInfraManager::EventCatcher::Runner#start_event_monitor) EMS [10.21.254.18] as [manageiq] Starting Event Monitor Thread
[----] I, [2019-08-15T07:12:12.921031 #8541:478bac8]  INFO -- : MIQ(ManageIQ::Providers::Redfish::PhysicalInfraManager#with_provider_connection) Connecting through ManageIQ::Providers::Redfish::PhysicalInfraManager: [hp01]
[----] E, [2019-08-15T07:12:13.743776 #8541:478bac8] ERROR -- : MIQ(ManageIQ::Providers::Redfish::PhysicalInfraManager::EventCatcher::Runner#start_event_monitor) EMS [10.21.254.18] as [manageiq] Event Monitor Thread aborted because [undefined method `listen' for nil:NilClass]
[----] E, [2019-08-15T07:12:13.743867 #8541:478bac8] ERROR -- : [NoMethodError]: undefined method `listen' for nil:NilClass  Method:[block (2 levels) in <class:LogProxy>]
[----] E, [2019-08-15T07:12:13.743897 #8541:478bac8] ERROR -- : /usr/local/lib/ruby/gems/2.5.0/bundler/gems/manageiq-providers-redfish-c8c92b27d884/app/models/manageiq/providers/redfish/physical_infra_manager/event_catcher/runner.rb:5:in `monitor_events'
/var/www/miq/vmdb/app/models/manageiq/providers/base_manager/event_catcher/runner.rb:161:in `block in start_event_monitor'
```

in the iLO, I have found some other hints. After adding manageiq, I can see some errors, that login attempts were failing and something between 5 and 10 parallel login sessions.

Is it possible, that manageiq is flodding the server somehow?

**Comment by tadeboro** [comment]
> Is it possible, that manageiq is flodding the server somehow?

Oh cr**, it most probably is. It is the event catcher component that is a bit too eager and tries to establish a SSE connection. And because iLO5 does not support SSE, the connection fails and MIQ just tries again and again and again ...

We can quickly test this hypothesis if you ssh into the MIQ appliance and do this:

 1. Run `vmdb` command. This will `cd` you into the MIQ's root folder on appliance.
 2. Open `lib/workers/miq_worker_types.rb` file, delete two lines that contain `ManageIQ::Providers::Redfish::PhysicalInfraManager::EventCatcher`, and save the changes.
 3. Run `appliance_console` command and press any key at initial prompt.
 4. Type in `14` (Stop EVM Server Processes) and confirm the operation. Wait for the operation to finish.
 5. When back in the main menu, type in `15` (Start EVM Server Processes) and confirm. Wait for the operation to finish.
 6. When back to the main menu, type in `19` (Quit).

Now that the event catcher is disabled, MIQ should stop spamming the BMC and things should start working a bit more reliably.

**Comment by daniel-wtd** [comment]
@tadeboro this... is... awesome!

It really happend to work and I was able to integrate and install 3 HPE Servers. The MACAddress and network information is still not shown in the webinterface, but creating the customization and pushing out the correct image was successful.

Thank you so much!

**Comment by tadeboro** [comment]
I am glad to hear that things started working.

What we did here, unfortunately, is not the proper fix for this problem. By removing one source of requests to the BMC, we improved the odds of not sending two or more concurrent requests to the BMC and thus reduced the chance of failure. But there are still other situations where the concurrent requests can be sent to the BMC, for example, if we start the provisioning process while the MIQ is performing an inventory refresh. And in such cases, there is still a significant chance of things falling apart.

I will create a separate issue about the event catcher. It would not hurt if we shut it down for BMCs that do not support SSE.

As for the display of the MAC addresses, I am pretty sure we do not store them during the inventory refresh process simply because we do not need it. But if you have a valid use-case that would make sysadmin's life more comfortable in any way, feel free to create another issue/feature request and we will see what we can do about this.
